File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.py
from typing import *
from collections import deque

VEC = ((-1, 0), (1, 0), (0, 1), (0, -1))
LIMIT = 9
if __name__ == '__main__':
    def bfs(_Q: Deque[Tuple[int, int]], _visited: List[List[bool]]) -> None:
        while _Q:
            y, x = _Q.popleft()

            for dy, dx in VEC:
                ny, nx = y + dy, x + dx
                if 0 <= ny < n and 0 <= nx < m:
                    if not _visited[ny][nx] and water[ny][nx] > 0:
                        _Q.append((ny, nx))
                        _visited[ny][nx] = True
                    if 0 < y < n - 1 and 0 < x < m - 1:
                        if water[ny][nx] == 0 and pool[ny][nx] <= pool[y][x]:
                            water[y][x] = 0
                        elif water[y][x] + pool[y][x] > water[ny][nx] + pool[ny][nx]:
                            if pool[ny][nx] < pool[y][x] and not water[ny][nx] or \
                                    pool[y][x] > water[ny][nx] + pool[ny][nx]:
                                water[y][x] = 0
                            elif pool[ny][nx] < pool[y][x]:
                                water[y][x] = min(water[y][x], abs(pool[ny][nx] + water[ny][nx] - pool[y][x]))
                            else:
                                water[y][x] = min(water[y][x], abs(pool[ny][nx] - pool[y][x]) + water[ny][nx])

    n, m = list(map(int, input().split()))
    pool: List[List[int]] = [[0] * m for _ in range(n)]
    water: List[List[int]] = [[0] * m for _ in range(n)]
    max_height: int = 0
    Q: Deque[Tuple[int, int]] = deque()
    visited: List[List[bool]] = [[False] * m for _ in range(n)]

    for i in range(n):
        line = list(map(int, list(input())))
        for j in range(m):
            pool[i][j] = line[j]
            max_height = max(max_height, pool[i][j])
            # Visit edges first
            if i == 0 or j == 0 or i == n - 1 or j == m - 1:
                Q.append((i, j))
                visited[i][j] = True
                water[i][j] = 0

    # Fill water first
    for i in range(1, n - 1):
        for j in range(1, m - 1):
            water[i][j] = max_height - pool[i][j]

    bfs(Q, visited)

    is_leftover_exist = True
    while is_leftover_exist:
        is_leftover_exist = False
        Q.clear()
        # Fix leftovers
        for i in range(1, n - 1):
            for j in range(1, m - 1):
                if water[i][j] == 0:
                    continue
                for di, dj in VEC:
                    ni, nj = i + di, j + dj
                    if pool[i][j] + water[i][j] > pool[ni][nj] + water[ni][nj]:
                        is_leftover_exist = True
                        Q.append((i, j))
                        break
        bfs(Q, visited)

    print(sum(sum(w) for w in water))

File: 백준/Gold/1113. 수영장 만들기/수영장 만들기.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_HEIGHT 9 + 1
#define MAX_POOL_WIDTH 50 + 1
#define VEC_LEN 4

typedef struct _node
{
    int y;
    int x;
    struct _node *next;
} node;

typedef struct _queue
{
    node *head;
    node *tail;
    int count;
} queue;

typedef enum
{
    false,
    true
} bool;

int VEC[][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

int n, m;
int pool[MAX_POOL_WIDTH][MAX_POOL_WIDTH];
bool visited[MAX_HEIGHT][MAX_POOL_WIDTH][MAX_POOL_WIDTH];
int edges[MAX_POOL_WIDTH * 4][2];
int water_filled;

void init_queue(queue *q)
{
    q->count = 0;
    q->head = NULL;
    q->tail = NULL;
}
void init_node(node *node, int y, int x)
{
    node->y = y;
    node->x = x;
    node->next = NULL;
}
void enque(queue *q, int y, int x)
{
    node *new_node = (node *)malloc(sizeof(node *));
    init_node(new_node, y, x);
    if (!q->count)
    {
        q->head = new_node;
    }
    else
    {
        q->tail->next = new_node;
    }
    q->tail = new_node;
    q->count++;
}
void deque(int *y, int *x, queue *q)
{
    if (!q->count)
    {
        return;
    }
    node *tmp = q->head;
    *y = q->head->y;
    *x = q->head->x;
    q->head = q->head->next;
    free(tmp);
    q->count--;
}

void bfs(queue *q, int level)
{
    while (q->count)
    {
        int y, x;
        deque(&y, &x, q);

        for (int i = 0; i < VEC_LEN; i++)
        {
            int ny, nx;
            ny = VEC[i][0] + y;
            nx = VEC[i][1] + x;

            if (0 < ny && ny < n - 1 && 0 < nx && nx < m - 1 && !visited[level][ny][nx]
                && ((pool[y][x] < pool[ny][nx] && level > pool[ny][nx])
                    || (pool[y][x] > pool[ny][nx] && pool[y][x] < level)
                    || (pool[y][x] == pool[ny][nx] && pool[y][x] < level)))
            {
                visited[level][ny][nx] = true;
                water_filled--;
                enque(q, ny, nx);
            }
        }
    }
}
int main()
{
    int edges_idx = 0;
    queue *q = (queue *)malloc(sizeof(queue *));
    init_queue(q);

    scanf("%d %d", &n, &m);

    for (int i = 0; i < n; i++)
    {
        char line[MAX_POOL_WIDTH];
        scanf("%s", line);
        for (int j = 0; j < m; j++)
        {
            pool[i][j] = line[j] - '0';
            // Fill water
            if (0 < i && i < n - 1 && 0 < j && j < m - 1)
            {
                water_filled += (MAX_HEIGHT - 1) - pool[i][j];
            }
            else
            {
                edges[edges_idx][0] = i;
                edges[edges_idx][1] = j;
                edges_idx++;
            }
        }
    }

    for (int level = MAX_HEIGHT - 1; level > 1; level--)
    {
        for (int i = 0; i < edges_idx; i++)
        {
            enque(q, edges[i][0], edges[i][1]);
        }
        bfs(q, level);
    }

    printf("%d\n", water_filled);
    return 0;
}
